### PR TITLE
hamt: preserve() traversal_stack uses rust stack

### PR DESCRIPTION
**Pull request by drbeefsupreme**
This was using the lightweight stack, but this presented the following issue: `NockStack::copy()` can be called during the `Hamt::preserve()` call, and the lightweight stack may not be empty. `NockStack::copy()` presumes that the lightweight stack is empty when it is called, and thus tried to pop elements that `preserve()` had pushed onto it.

`NockStack::copy()` should probably assert that the lightweight stack is empty when its called.

I'm making the assert `stack_is_empty()` a separate PR #85 since its a separate concern, but I have tested that it works fine with this PR and several jamfiles where `copy()` is called many times.

### CODE REVIEW

**Comment by drbeefsupreme** [comment]
Hold off on this one moment... I thought I tested this with my jet cold state branch, but I must have made a mistake somewhere since now its not working.

**Comment by drbeefsupreme** [comment]
Nevermind, all good. I just forgot that I had some of the changes in the jet cold state branch I needed stashed, not committed, and its working with the stash restored.

**Comment by ashelkovnykov** [comment]
Haven't looked at the code in this PR yet, but I just found that `%memo` hints were causing Ares crashes on `preserve` calls to the cache in my stack trace test pill. As another data point, I'm going to try applying this PR to my branch to see if it fixes the crashes.

File: rust/ares/src/hamt.rs [hunk 1]
```diff
@@ -437,17 +437,26 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
...
+            traversal_stack[0] = Some((self.0, 0));
             let mut traversal_depth = 1;
             'preserve: loop {
                 if traversal_depth == 0 {
                     break;
                 }
-                let (stem, mut position) = *(stack.top::<(Stem<T>, u32)>());
+                let Some((stem, mut position)) = traversal_stack[traversal_depth - 1] else {
+                    panic!("Attempted to access uninitialized array element");
+                };
```
Comment (eamsden) [inline rust/ares/src/hamt.rs:456]:
```suggestion
                let (stem, mut position) = traversal_stack[traversal_depth - 1].expect("Attempted to access uninitialized array element");
```

`.expect()` is the canonical way to do this for both `Result` and `Option`

**Review by eamsden — CHANGES_REQUESTED** [review]
Just one change for idiomatic/succinct Rust.

**Comment by ashelkovnykov** [comment]
> Haven't looked at the code in this PR yet, but I just found that `%memo` hints were causing Ares crashes on `preserve` calls to the cache in my stack trace test pill. As another data point, I'm going to try applying this PR to my branch to see if it fixes the crashes.

Confirmed that this PR fixes `%memo` hints in my stack traces pill

**Review by eamsden — APPROVED** [review]

**Review by ashelkovnykov — COMMENTED** [review]
Late, but LGTM.

The code all looks logically correct, though I don't understand the Ares `Hamt` code fully (for lack of trying).

**Pull request merged**